src/object/tcb.c

```diff
@@ -792,6 +792,48 @@ static exception_t decodeSetTLSBase(cap_t cap, word_t length, word_t *buffer)
     return invokeSetTLSBase(TCB_PTR(cap_thread_cap_get_capTCBPtr(cap)), tls_base);
 }
 
+static void invokeSetFlags(tcb_t *thread, word_t clear, word_t set, bool_t call)
+{
+    word_t flags = thread->tcbFlags;
+
+    flags &= ~clear;
+    flags |= set & seL4_TCBFlag_MASK;
+    thread->tcbFlags = flags;
+
+#ifdef CONFIG_HAVE_FPU
+    /* Save current FPU state before disabling FPU: */
+    if (set & seL4_TCBFlag_fpuDisabled) {
+        fpuRelease(thread);
+    }
+#endif
+    if (call) {
+        word_t *ipcBuffer = lookupIPCBuffer(true, thread);
+        setRegister(thread, badgeRegister, 0);
+        unsigned int length = setMR(thread, ipcBuffer, 0, flags);
+        setRegister(thread, msgInfoRegister, wordFromMessageInfo(
+                        seL4_MessageInfo_new(0, 0, 0, length)));
+    }
+    setThreadState(thread, ThreadState_Running);
```

Comment:
Is setting to `ThreadState_Running` needed or not? `handleInvocation` seems to change restart to running in the normal case, but it's behind an `unlikely`?

Reply:
I haven't dug through the code to work out if it's actually needed but from the verification perspective, this pattern with the `setThreadState ThreadState_Running` is one that we've already verified elsewhere. This means that unless there is a specific need to avoid the `setThreadState ` that we would have a strong preference to keep it as is so that we can reuse those proofs.

Reply:
But I only added it recently and you didn't complain about it missing before. Most calls don't seem to do the `setThreadState ThreadState_Running`, but some do and I don't see a pattern.

Reply:
Our proofs require it as part of the reply from kernel pattern, so I didn't previously complain about it because it wasn't needed until now. I haven't done an exhaustive search, but you should see a `setThreadState(NODE_STATE(ksCurThread), ThreadState_Running)` in all of the verified invocations that are returning something to the user.